### PR TITLE
4.x typehints 3

### PR DESCRIPTION
**Pull request by ADmad**
Update packages `Error` to `Http`.

### CODE REVIEW

File: src/Http/Cookie/Cookie.php [hunk 1]
```diff
@@ -236,10 +236,22 @@ public function getValue()
...
+     * Gets the cookie value as a string.
+     *
+     * This will collapse any complex data in the cookie with json_encode()
+     *
+     * @return mixed
+     * @deprecated 4.0.0 Use getScalarValue() instead.
      */
     public function getStringValue()
+    {
+        return $this->getScalarValue();
```
Comment (inoas) [inline src/Http/Cookie/Cookie.php]: `return (string)$this->getScalarValue();`?

aka could this not be an integer? The method name suggests otherwise.

Comment (ADmad) [reply src/Http/Cookie/Cookie.php]: The old method name was misleading, hence it's deprecated and new method added instead.

File: src/Http/Cookie/Cookie.php [hunk 2]
```diff
@@ -236,10 +236,22 @@ public function getValue()
...
 
     /**
-     * {@inheritDoc}
-     * @psalm-suppress InvalidReturnType
+     * Gets the cookie value as a string.
+     *
+     * This will collapse any complex data in the cookie with json_encode()
+     *
+     * @return mixed
+     * @deprecated 4.0.0 Use getScalarValue() instead.
```
Comment (markstory) [inline src/Http/Cookie/Cookie.php]: When is the return not a string?

Comment (ADmad) [reply src/Http/Cookie/Cookie.php]: Whenever it's not an array :)

Assertions in `ResponseTest::testWithCookieScalar()` (which I haven't changed) fail if the return value for the method is type hinted as `string`.

File: src/Http/Cookie/CookieInterface.php [hunk 3]
```diff
@@ -49,13 +49,13 @@ public function getName(): string;
...
-     * Gets the cookie value as a string.
+     * Gets the cookie value as scalar.
      *
      * This will collapse any complex data in the cookie with json_encode()
      *
-     * @return string
+     * @return mixed
      */
-    public function getStringValue();
+    public function getScalarValue();
```
Comment (dereuromark) [inline src/Http/Cookie/CookieInterface.php]: Do we need to document this in the migration guide etc?

File: src/Http/ServerRequest.php [hunk 4]
```diff
@@ -380,12 +380,16 @@ protected function _processGet(array $query, string $queryString = ''): array
     /**
      * Process uploaded files and move things onto the post data.
      *
-     * @param array $post Post data to merge files onto.
+     * @param mixed $post Post data to merge files onto.
      * @param array $files Uploaded files to merge in.
      * @return array merged post + file data.
      */
     protected function _processFiles($post, array $files)
```
Comment (dereuromark) [inline src/Http/ServerRequest.php]: When can this ever not operate on array? I checked, the code before for sure returns an array

    $config['post'] = $this->_processPost($config['post']);

As such I wonder if we should even allow for this case here and weaken the API?

File: src/Http/ServerRequest.php [hunk 5]
```diff
@@ -1881,7 +1885,7 @@ public function getUri()
      * @param bool $preserveHost Whether or not the host should be retained.
      * @return static
      */
-    public function withUri(UriInterface $uri, $preserveHost = false)
+    public function withUri($uri, $preserveHost = false)
```
Comment (dereuromark) [inline src/Http/ServerRequest.php]: Could this be kept? Also, bool addition possible?

**Comment by dereuromark** [comment]
Rest looks good to me.

File: src/Http/Cookie/CookieInterface.php (hunk 3 above)
Comment (ADmad) [reply src/Http/Cookie/CookieInterface.php]: The old method still exists with deprecated tag but yeah mentioning it in migration guide would also be nice.

File: src/Http/ServerRequest.php (hunk 5 above)
Comment (ADmad) [reply src/Http/ServerRequest.php]: These methods are inherited from the PSR interface so can't add additional typehints for arguments.

File: src/Http/ServerRequest.php (hunk 4 above)
Comment (ADmad) [reply src/Http/ServerRequest.php]: There's a test case where POST data is string:

https://github.com/cakephp/cakephp/blob/b1673c25d8955cc69e551fa97e7a159bb335f82a/tests/TestCase/Http/ServerRequestTest.php#L2437-L2443

I can change the `mixed` to `array|string` if you prefer.

File: src/Http/ServerRequest.php (hunk 5 above)
Comment (ADmad) [reply src/Http/ServerRequest.php]: My bad, the interface method does have the `UriInterface` typehint.

File: src/Http/ServerRequest.php (hunk 4 above)
Comment (dereuromark) [reply src/Http/ServerRequest.php]: sounds fair.

File: src/Http/ServerRequest.php (hunk 5 above)
Comment (dereuromark) [reply src/Http/ServerRequest.php]: What about the 2nd arg? bool typehint?

**Pull request merged**